### ISSUE TITLE
Improve ARIA attributes of gutter and scroller

### ISSUE THREAD

**Issue by szebniok** [ajaxorg/ace#5788]
### Describe the bug

When keyboard accessibility mode is turned on, users can tab through two items: gutter and scroller. The current ARIA labels for these elements are not optimal for screen reader users. The labels are overloaded with instructions about available keyboard shortcuts, which would be better suited as descriptions.

### Expected Behavior

Screen reader users should be able to understand the context of the gutter and scroller controls. The instructions should be moved to descriptions that are usually read after a short pause. [Excerpt from WAI-ARIA spec.](https://www.w3.org/TR/wai-aria/#:~:text=There%20are%20not%20any%20WAI%2DARIA%20widget%20roles%20that%20correspond%20to%20such%20an%20interaction%20model%20so%20an%20author%20could%20give%20the%20slide%20container%20role%20application%2C%20an%20aria%2Droledescription%20of%20%22Slide%20Editor%22%2C%20and%20use%20aria%2Ddescribedby%20to%20provide%20instructions.)

| control  | label                                 | description                                                                  |
|----------|---------------------------------------|------------------------------------------------------------------------------|
| Gutter   | `${textInputAriaLabel}`, Editor gutter    | Press Enter to interact with controls using arrow keys, press Escape to exit |
| Scroller | `${textInputAriaLabel}`, Editor content   | Press Enter to start editing, press Escape to exit                           |
| Editor   | `${textInputAriaLabel}`, Cursor at row `$0` |                                                                              |

### Current Behavior

- Gutter has an `aria-label` set to `"Editor gutter, press Enter to interact with controls using arrow keys, press Escape to exit"`.
- Scroller has an `aria-label` set to `"Editor content, press Enter to start editing, press Escape to exit"`.
- `textInputAriaLabel` is only announced once the user interacts with the scroller by pressing the Enter key.

| control  | label                                                                                       | description |
|----------|---------------------------------------------------------------------------------------------|-------------|
| Gutter   | Editor gutter, press Enter to interact with controls using arrow keys, press Escape to exit |             |
| Scroller | Editor content, press Enter to start editing, press Escape to exit                          |             |
| Editor   | `${textInputAriaLabel}`, Cursor at row `$0`                                                       |             |

### Reproduction Steps

1. Turn on keyboard accessibility mode.
2. Observe the ARIA labels for gutter and scroller.

### Possible Solution

1. Extract instructions to two [visually hidden](https://www.a11yproject.com/posts/how-to-hide-content/) elements which are linked to scroller and gutter via `aria-describedby`.
2. Prepend scroller’s and gutter’s `aria-label` with `textInputAriaLabel`.

### Additional Information/Context

Drawbacks:
- Nls strings `editor.gutter.aria-label` and `editor.scroller.aria-label` will be changed.
- Two visually hidden HTML elements need to be created since `aria-description` is not yet part of the ARIA spec.

I am willing to contribute this change once I receive positive feedback.

### Ace Version / Browser / OS / Keyboard layout

v1.39.1 / Chrome / Mac / VoiceOver

**Comment by marinsokol5** [comment]
Hey, 
First of all, great issue description, thanks for the feedback and the context! 

Regarding moving usage instructions from `aria-label` to `aria-describedby`, that does sounds very reasonable. 
We would just need to test if the solution works across combination of screen reader (NVDA, VoiceOver, ...) + OS + browser, cause sometimes the screen reader would skip `aria-describedby` when `aria-label` is there. 
I am not sure about the benefit of it though, since if we expect the screen reader to just read it a bit later (after a small pause), it sounds like it's the same experience in the end as when these instructions are part of the `aria-label` itself. 
I understand it might be "more according to the spec", but keeping them in `aria-label` does ensure they are being read out. 
If you are willing to contribute a PR, we can test it on our end and help with merging!

Regarding prepending `textInputAriaLabel`, that also sounds reasonable but it should be also possible today by providing a different string for these keys in ace.config -> text-input.aria-label, editor.gutter.aria-label, editor.scroller.aria-label. 
```
const editorLabel = "My cool JSON editor";
ace.config.setMessages({
     ....
    "text-input.aria-label": `${editorLabel}, Cursor at row $0`,
    "editor.scroller.aria-label": `${editorLabel} content, press Enter to start editing, press Escape to exit`,
    "editor.gutter.aria-label": `${editorLabel} gutter, press Enter to interact with controls using arrow keys, press Escape to exit`,
 })
```
You can also from consumer side have a container element around Ace editor with a specific ariaLabel (such as that `editorLabel` from above, and with `role='group'` that would be focusable and tabbable into, and then because it's a group that aria-label will be prepended to it's children focusable elements, such as gutter, scroller and text-input. Although I have to say that based on past experiences, the group functionality can also be a bit flaky in MacOS + Voiceover, so I think the first approach is more reliable.

**Comment by szebniok** [comment]
Hi @marinsokol5,

Following up, I tested moving the instructions to `aria-describedby`. As you suspected, the experience was inconsistent: the descriptions weren't announced by default in Chrome and Safari. They _did_ work in Firefox. I was a bit suprised by that, as `aria-describedby` seems to work reliably for other types of inputs.

Given that this approach doesn't work reliably, it doesn't seem like the best path forward. I'll close this issue.

Thanks again for the helpful discussion!
